File: 716-max-stack/716-max-stack.java
class MaxStack {
    Stack<Integer> stack, maxStack;

    public MaxStack() {
        stack = new Stack<>();
        maxStack = new Stack<>();
    }

    public void push(int x) {
        stack.push(x);
        if (maxStack.isEmpty()) 
            maxStack.push(x); 
        else {
            maxStack.push(Math.max(x,maxStack.peek()));
        }
    }

    public int pop() {
        int ret = stack.pop();
        maxStack.pop();

        return ret;
    }

    public int top() {
        return stack.peek();
    }

    public int peekMax() {
        return maxStack.peek();
    }

    public int popMax() {
        int max = maxStack.peek();
        Stack<Integer> buffer = new Stack<>();
        while (stack.peek() != max) {
            buffer.push(pop());
        }
        pop();
        while (!buffer.isEmpty()) 
            push(buffer.pop());

        return max;
    }
}
/**
 * Your MaxStack object will be instantiated and called as such:
 * MaxStack obj = new MaxStack();
 * obj.push(x);
 * int param_2 = obj.pop();
 * int param_3 = obj.top();
 * int param_4 = obj.peekMax();
 * int param_5 = obj.popMax();
 */

File: LinkedIn/716-max-stack/716-max-stack.cpp
class MaxStack {
    map<int, vector<list<int>::iterator>>m;
    list<int>dl;
public:
    MaxStack() {
        
    }
    
    void push(int x) {
        dl.push_back(x);
        m[x].push_back(--dl.end());
    }
    
    int pop() {
        int top = this->top();
        dl.pop_back();
        m[top].pop_back();
        if(m[top].size()==0)
            m.erase(top);
        return top;
    }
    
    int top() {
        return *(dl.rbegin());
    }
    
    int peekMax() {
        return m.rbegin()->first;
    }
    
    int popMax() {
        int max = peekMax();
        auto it = m[max].back();
        dl.erase(it);
        m[max].pop_back();
        if(m[max].size()==0)
            m.erase(max);
        return max;
    }
};

/**
 * Your MaxStack object will be instantiated and called as such:
 * MaxStack* obj = new MaxStack();
 * obj->push(x);
 * int param_2 = obj->pop();
 * int param_3 = obj->top();
 * int param_4 = obj->peekMax();
 * int param_5 = obj->popMax();
 */